ir/CMakeLists.txt

```diff
@@ -69,6 +69,13 @@ set (IR_DEF_FILES ${IR_DEF_FILES} ${BASE_IR_DEF_FILES} PARENT_SCOPE)
 
 add_library (ir STATIC ${IR_SRCS})
 target_link_libraries(ir PRIVATE absl::flat_hash_map ${LIBGC_LIBRARIES})
+
+# We need this flag to allow delete operations with "size" parameters.
+# GCC enables this flag by default but Clang does not.
```

Comment:
I think we need to do the opposite thing: guard sized `delete` with feature check. Note that `-fsized-deallocation` is an ABI break, so it might not be appropriate for some downstream users.

Reply:
It looks like this is a C++14 feature but Clang has neglected to implement it:
https://reviews.llvm.org/D112921

So it should definitely be the default behavior. We could add a feature flag for folks that want to keep it off but I am not sure why they would want that. 

Maybe we do not even need to add a sized delete and a simple delete is sufficient: 
https://developers.redhat.com/blog/2021/05/05/detecting-memory-management-bugs-with-gcc-11-part-2-deallocation-functions#mismatches_between_new_and_delete_operators

I do not have GCC 11 at hand.

Reply:
This mostly comes does to what C++ version are we using?  We'd previously made a decision to try to converge on C++20, so we should use whatever flags are needed to get the compiler to support C++20 if at all possible.  If we find compilers that really cannot support C++20, then we can put in checks/workarounds in the code for those compilers.

We should be detecting which versions of clang need this flag to support C++20, and only apply it then.

Reply:
I do not see why we need to do any version detection, note that apple clang and mainline clang are different clangs, and any version checks are fragile by definition. Even more, sized deallocation could be a platform choice (Apple platforms including).

We can just check via `#ifdef __cpp_sized_deallocation` and provide sized delete's only when the feature is enabled.

Reply:
If platforms that follow C++20 consistently set `__cpp_sized_deallocation` and those that don't follow C++20 don't set it, that would be fine.  But I can't find any mention of this preprocessor macro in any version of the C++ spec.

Reply:
These are feature testing macro and they are essentially implementation defined: https://en.cppreference.com/w/cpp/feature_test

FWIW: clang recently enabled `-fsized-deallocation` by default: https://github.com/llvm/llvm-project/pull/90373 so this is part of LLVM 19 release and further on. Likely this will be default in Apple fork in 1-2 years, however, it will certainly be SDK version-specific (e.g. enabling compatibility with lower SDK versions would default to `-fnosized-deallocation` and newer ones will be `-fsized-deallocation`) due to ABI compatibility

One example of check could be seen at https://github.com/google/cel-cpp/pull/558/files:
```diff
-      ::operator delete(static_cast<void*>(p), n * sizeof(T),
+#if defined(__cpp_sized_deallocation) && __cpp_sized_deallocation >= 201309L
+      ::operator delete(static_cast<void *>(p), n * sizeof(T),
+                        static_cast<std::align_val_t>(alignof(T)));
+#else
+      ::operator delete(static_cast<void *>(p),
+                        static_cast<std::align_val_t>(alignof(T)));
+      static_cast<void>(n); // unused
+#endif
```

Similar code is used on Boost to make it compatible (https://www.boost.org/doc/libs/1_84_0/boost/container/new_allocator.hpp), jemalloc, libc++, etc.

Reply:
I agree that `__cpp_sized_deallocation` should be the way to go unless it is broken in some of the compilers which we target. This should be much more robust than checking compiler version.